### DISCUSSION TITLE
Export clusters identified by FlowSOM

### DISCUSSION THREAD

**Discussion by ReubenMcG** [ImmuneDynamics/Spectre#75] [Help] [answered]
Hi all,

First off, thank you for a great package, I have found Spectre in R so far very user friendly and I have just skimmed the surface.

I have a couple of questions and please feel free to point me to relevant tutorials or resources that I may have overlooked.

1)I have followed the **Simple discovery workflow** and was wondering if there was a way to check dot plots of various cluster identified by FlowSOM. Say we think ,based on the heat map, a given cluster is T regulatory cells and want to check if this is true based on classical 2D dot plots, either by exporting the data in a way that can be read by FlowJo (too look at back gating), or is there a way of doing this in R directly.

In general I am talking about validating identified cluster by classical 2D dot plot methods, how would you recommend going about this? I am envisaging exporting the dataset as an FCS files with annotated clusters included, and then importing to flowjo to check expression of markers on 2D plots.

2) Is there a way of visualising manually gated populations from FlowJo on the dimensionality reduced plots (uMAP or tSNE plots) within R using Spectre. Taking the example above again, if I had gated on Tregs on Flowjo and want to see where these cells are located in a uMAP produced in Spectre.

Thanks

Reuben

**Comment by tomashhurst** [comment, marked as answer]
Hi @ReubenMcG, thanks so much!

**"...if there was a way to check dot plots of various cluster identified by FlowSOM"**

Definitely! Either you can make some plots of the marker of interest directly (e.g. rather than making a plot of UMAP_X vs UMAP_Y coloured by cluster/population, you could make one of CD3_asinh vs FoxP3_asinh or similar. 

Here is an example with the demo data:

```
library('Spectre')
cell.dat <- Spectre::demo.clustered
make.colour.plot(dat = cell.dat, x.axis = 'CD11b_asinh', y.axis = 'CD45_asinh', col.axis = 'Population', add.label = TRUE)
```

![Factor plot - Population - plotted on CD11b_asinh by CD45_asinh](https://user-images.githubusercontent.com/11766139/127079449-b185345b-9529-4b04-a2c0-1a6be1890537.png)

However, to get into more detail you can easily produce FCS files and pull them into FlowJo. In the 'annotation' section there is a command there that generates FCS files, and you can pull them directly into FlowJo! All the raw and asinh columns will be in there, as will be the cluster and population data -- they'll just be converted to numbers, but it's not too difficult to figure out which is which.

![image](https://user-images.githubusercontent.com/11766139/127079476-9aabe7df-6e58-42d1-905c-d194ba09ab38.png)

My typical method for this is to pull the files back into FlowJo.

**"Is there a way of visualising manually gated populations from FlowJo on the dimensionality reduced plots (uMAP or tSNE plots) within R using Spectre"

Indeed there is! In the simple discovery script the FCS files are generate from the 'full' dataset, but you can easily do this with the subsampled dataset (with the UMAP/tSNE etc) as well.

Here is the existing code for the full dataset:

```
    ### Write FCS files
        
        setwd(OutputDirectory)
        setwd("Output 3 - annotation")
        
        dir.create('FCS files')
        setwd('FCS files')
        
        write.files(cell.dat,
                    file.prefix = exp.name,
                    divide.by = sample.col,
                    write.csv = FALSE,
                    write.fcs = TRUE)
```

You could then run this to do the same for the subsampled data but running the `write.files` function on `cell.sub`:

```
    ### Write FCS files
        
        setwd(OutputDirectory)
        setwd("Output 3 - annotation")
        
        dir.create('FCS files - subsampled')
        setwd('FCS files - subsampled')
        
        write.files(cell.sub,
                    file.prefix = exp.name,
                    divide.by = sample.col,
                    write.csv = FALSE,
                    write.fcs = TRUE)
```

**Reply by ReubenMcG** [reply]
Thanks so much,

I am beginning to understand more and more now the advantage of the data.table structure. And the advantages of having an immunologist like yourself helping develop this package. And sorry I think I could have got this answer by working further through the workflow to the write.fcs example!

I think question 1) is answered in both parts of your answer thanks.

I may not have explained Q2 properly.

2)Essentially I have clustered data generated the uMAPS as per workflow and think I have identified Tregs based on the expression heat map etc. But to be 100% sure I would like to see wether the Treg population I have gated manually in FlowJo does indeed overlay with the clusters I have merged and annotated in Spectre... 

If I have used FlwoJo to do traditional gating and gate on Tregs, then exported the Treg population as a csv (as explained in other sections of the workflow for Spectre). I then would like to see where this manually gated population sits on uMAP or tSNE generated in spectre?

Hopefully this makes more sense? And thank you so much for taking the time, really appreciated

**Reply by tomashhurst** [reply]
@ReubenMcG sorry for the terribly late reply! I think I see what you are getting at. If you take the FCS files created at the end of the Spectre workflow back into FlowJo. These should have the UMAP coordinates and the cluster numbers embedded. You can re-draw your typical gates (e.g. CD3+ --> CD4+ --> CD25+CD127lo --> FoxP3+) on those files, and then create an overlay in the layout editor -- gated T regs overlaid on all cells, with UMAP_X and UMAP_Y as the X and Y axis. You should then be able to see where the manually gated T regs sit on the UMAP plot. Additionally, you can see how well the clusters align. For example, let's say your T regs are in metaclusters 2 and 5 -- if you gate the T regs and then put the FlowSOM_metacluster on the axis, you should see all the cells stacking up at 2 and 5. I _think_ that's an answer to your question, but if I have mis-understood then please do correct me!

**Reply by ReubenMcG** [reply]
@tomashhurst yes that is exactly what I meant. 

Thank you so much again. All coming together nicely now. Looking forward to several of the "COMING SOON" workflows, particularly the "Multi-level analysis for profiling of rare populations" and "Advanced quantitative and statistical analysis". For now I have sub-gated in Spectre on various cell types (t-cell, b-cell etc) and re-done the basic discovery workflow to sub cluster more rare  populations within those cluster. Then pooled the stats into one big table for PCA etc, seems to work quite well.

**Reply by SamGG** [reply]
Working in 2 (or more) successive levels/gating steps is usually required as there are so many cell types addressed in the same sample. Moreover, as clustering puts all clusters (aka cell populations) on the same unique level. The percentage of cells are referenced to the total,  as the classical cell hierarchy is lost. This is a problem when the percentage to the parent or the grand-parent might be of interest.

**Reply by ReubenMcG** [reply]
@SamGG thanks, makes sense. I am interested in say cases vs controls and differences between them. So I guess the percentage as in classical gating or hierarchy vs entire population matters too much.

**Reply by tomashhurst** [reply]
The multi-level workflow and advanced stats one are in use for a couple of our active studies, but I haven't had the chance to refine them yet (i.e. they are still full of weird things that would only make sense to me). Those studies are wrapping up soon so I'll take the chance then to get them online!
